File: app/src/main/java/com/saienko/androidthings/barman/ui/settings/adapter/SelectComponentAdapter.java
package com.saienko.androidthings.barman.ui.settings.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;
import com.saienko.androidthings.barman.R;
import com.saienko.androidthings.barman.db.cocktail.Component;
import com.saienko.androidthings.barman.ui.settings.adapter.holder.SpinnerComponentViewHolder;

import java.util.ArrayList;
import java.util.List;

/**
 * Created
 * User: Vasiliy Saienko
 * Date: 9/10/17
 * Time: 17:25
 */

public class SelectComponentAdapter extends ArrayAdapter<Component> {

    public SelectComponentAdapter(@NonNull Context context,
                                  List<Component> components) {
        super(context, R.layout.spinner_item_group, components);
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        SpinnerComponentViewHolder holder;
        Component                  item = getItem(position);

        if (convertView == null) {
            convertView = LayoutInflater.from(getContext()).inflate(R.layout.spinner_item_gpio, parent, false);
            holder = new SpinnerComponentViewHolder();
            holder.name = convertView.findViewById(R.id.motorName);
            convertView.setTag(holder);
        } else {
            holder = (SpinnerComponentViewHolder) convertView.getTag();
        }

        holder.name.setText(item.getName());
        return convertView;
    }

    @Override
    public View getDropDownView(int position, View convertView, @NonNull ViewGroup parent) {

        Component      item     = getItem(position);
        LayoutInflater inflater = LayoutInflater.from(getContext());
        View           row      = inflater.inflate(R.layout.spinner_item_gpio, parent, false);
        TextView       txtTitle = row.findViewById(R.id.motorName);

        txtTitle.setText(item.getName());
        return row;
    }

    public List<Component> getItems() {
        List<Component> list = new ArrayList<>(getCount());
        for (int i = 0; i < getCount(); i++) {
            list.add(getItem(i));
        }
        return list;
    }

}

File: app/src/main/java/com/saienko/androidthings/barman/ui/settings/adapter/SelectComponentAdapter.kt
package com.saienko.androidthings.barman.ui.settings.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.saienko.androidthings.barman.R
import com.saienko.androidthings.barman.db.entity.Component
import com.saienko.androidthings.barman.ui.settings.adapter.holder.SpinnerComponentViewHolder

/**
 * Created
 * User: Vasiliy Saienko
 * Date: 9/10/17
 * Time: 17:25
 */

class SelectComponentAdapter(context: Context,
                             components: List<Component>) : ArrayAdapter<Component>(context,
                                                                                    R.layout.spinner_item_group,
                                                                                    components) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val holder: SpinnerComponentViewHolder
        val item = getItem(position)

        if (convertView == null) {
            view = LayoutInflater.from(context).inflate(R.layout.spinner_item_component, parent, false)
            holder = SpinnerComponentViewHolder(view.findViewById(R.id.componentName))
            holder.name = view.findViewById(R.id.componentName)
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as SpinnerComponentViewHolder
        }

        holder.name.text = item.componentName
        return view
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {

        val item = getItem(position)
        val inflater = LayoutInflater.from(context)
        val row = inflater.inflate(R.layout.spinner_item_component, parent, false)
        val txtTitle = row.findViewById<TextView>(R.id.componentName)

        txtTitle.text = item.componentName
        return row
    }

}
